scipy/optimize/_differentiable_functions.py

```diff
@@ -0,0 +1,405 @@
+                    self.J = approx_derivative(fun_wrapped, self.x, f0=self.f,
+                                               **finite_diff_options).toarray()
+                    self.J_updated = True
+                    return self.J
+                self.J = self.J.toarray()
+                self.sparse_jacobian = False
+
+            else:
+                def evaluate_jac():
+                    if not self.f_updated:
+                        self.evaluate_fun()
+                    self.J = np.atleast_2d(
+                        approx_derivative(fun_wrapped, self.x, f0=self.f,
+                                          **finite_diff_options))
+                    self.J_updated = True
+                    return self.J
+                self.J = np.atleast_2d(self.J)
+                self.sparse_jacobian = False
+        self.evaluate_jac = evaluate_jac
+
+        # Define Hessian
+        if callable(hess):
+            self.H = hess(self.x, self.v)
+            self.H_updated = True
+
+            if sps.issparse(self.H):
+                def hess_wrapped(x, v):
+                    return sps.csr_matrix(hess(x))
+                self.H = sps.csr_matrix(self.H)
+
+            elif isinstance(self.H, LinearOperator):
+                def hess_wrapped(x, v):
+                    return hess(x, v)
+
+            else:
+                def hess_wrapped(x, v):
+                    return np.atleast_2d(np.asarray(hess(x, v)))
+                self.H = np.atleast_2d(np.asarray(self.H))
+
+            def evaluate_hess():
+                self.H = hess_wrapped(self.x, self.v)
+                self.H_updated = True
+                return self.H
+
+        elif hess in FD_METHODS:
+            def jac_dot_v(x, v):
+                return jac_wrapped(x).T.dot(v)
+            self.H = approx_derivative(jac_dot_v, self.x,
+                                       f0=self.J.T.dot(self.v),
+                                       args=(self.v,),
+                                       **finite_diff_options)
+            self.H_updated = True
+
+            def evaluate_hess():
+                if not self.J_updated:
+                    self.evaluate_jac()
+                self.H = approx_derivative(jac_dot_v, self.x,
+                                           f0=self.J.T.dot(self.v),
+                                           args=(self.v,),
+                                           **finite_diff_options)
+                self.H_updated = True
+                return self.H
+        elif isinstance(hess, HessianUpdateStrategy):
+            self.H = hess
+            self.H.initialize(len(x0), 'hess')
+            self.H_updated = True
+            self.J_prev = deepcopy(self.J)
+
+            def evaluate_hess():
+                if not self.J_updated:
+                    self.evaluate_jac()
+                delta_x = self.x - self.x_prev
+                delta_grad = self.J.T.dot(self.v) - self.J_prev.T.dot(self.v)
```

Comment:
Should `self.v_prev` be in the second term?

Reply:
It is right.  It took me some time to process this as well, but it should be ``self.v``in both terms.

For some reference in this, look in [Nocedal and Wright](http://www.bioinfo.org.cn/~wangchao/maa/Numerical_Optimization.pdf), p. 538, Eq. (18.20),
under the section "Full Quasi-Newton Approximations".